Auto-clean projects before compiling

elopez opened the pull request:
A common issue experienced by Slither users are crashes due to out of date artifacts, which often can be resolved by running a project clean. This adds a clean function to the multiple platforms to do be able to do this, and defaults to running it unless ignore_compile or skip_clean are enabled.

elopez commented:
I think the platform init kwargs don't contain all the keys, I'll have to revisit that bit 😅

bohendo commented:
I merged `dev` into this branch and fixed the `pylint` issue (oops and maybe I introduced a new `black` issue, I can fix this tomorrow).

Is there more work that needs to be done apart from getting CI passing or is the init kwargs thing separate? If you give me some pointers on what's left to do I'd be happy to help push this PR over the finish line because it's really important IMO, I've tripped over this issue many times.

elopez commented:
@bohendo thanks! I had some unpushed commits, so I rebased them & pushed for completeness. I think I addressed the kwargs concerns I had already. CI looks healthy (other than the random failures of brownie querying github, which should be helped with #342)

You'll see some hardhat oddness in the diff because master & dev have diverged. Once master is re-merged in dev and dev is merged back here, it should go away.

Pending tasks here, feel free to work on them and push here 👍 
* [ ] Adding cleanup commands for more platforms. Any remaining ones, we can tackle later on as well.
  - [ ] Brownie (I didn't see a way to clean)
  - [ ] Buidler, Embark, Etherlime, Waffle (probably not very relevant, haven't seen them in the wild lately?)
  - [ ] Etherscan (maybe we can add some cleanup here but it'd require manual code I think. I'm wary of writing "rm" code)
  - [ ] Truffle
  - [X] Hardhat, Dapp, Foundry
  - [X] solc, archive, standard (I don't think there's anything to clean here?)
* [X] Config option and CLI flag
* [ ] Some testing for this
* [ ] Review / cleanups if needed

Going forward, maybe we can refactor the platform classes, I feel it does not make much sense to pass kwargs on each method call instead of one-time during instantiation 🤔

bohendo commented:
Remaining platforms:
- Brownie: no cache but we could `rm -rf build`, I think this isn't relevant to the issue of stale cache + general sketchiness of automated `rm -rf` means we should prob skip this
- buidler: replaced by hardhat, I'm not too worried about maintaining pristine support for this one
- Embark & Etherlime: not updated in > a year, can probably be considered de facto depreciated even if they're not officially retired projects
- Waffle: wrapper around {js-}solc, don't think it's doing anything different than native compilation
- Etherscan: similar to brownie, I don't think the build artifacts generated will break anything
- Truffle: we're already providing the `--all` flag so that should force-override any stale artifacts

IMO, hardhat, foundry and dapp stand out as the ones that really need a cleanup step. If we ship auto-cleanup for these 3 then I'd consider this problem well resolved & can open new issues/PRs if we ever trip over stale artifacts on some other platform (but I suspect we won't..)

Re Testing.. what's the best way to reproduce this issue? I thought I'd compile, change a contract name, and then re-compile but that worked first try. I remember hitting this issue in the context of missing imports from old contracts that have been deleted/modified but now I'm wondering whether the [node modules resolution PR](https://github.com/crytic/crytic-compile/pull/324) fixed that issue & we'll have trouble setting up a failing test case now.. 🤔

elopez commented:
I tried a few things with hardhat but I've not been able to reproduce the stale artifacts, so I don't have a good, realistic test case for this 🙈 

Maybe we could create a "fake" artifact, run crytic-compile and see that it was removed? 🤔

montyly commented:
fun fact, we should not need this with hardhat because we run it with `--force`. I guess this flag does not work as expected 🤔
